sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java

```diff
@@ -1642,43 +1642,45 @@ public Object restriction() {
               .build());
     }
 
-    ByteString.Output primaryBytes = ByteString.newOutput();
-    ByteString.Output residualBytes = ByteString.newOutput();
-    try {
-      fullInputCoder.encode(windowedSplitResult.getPrimarySplitRoot(), primaryBytes);
-      fullInputCoder.encode(windowedSplitResult.getResidualSplitRoot(), residualBytes);
-    } catch (IOException e) {
-      throw new RuntimeException(e);
+    if (windowedSplitResult.getResidualSplitRoot() != null) {
```

Comment:
Can't `windowedSplitResult` itself be null here?

Reply:
`windowedSplitResult` can be null and it has been checked above https://github.com/apache/beam/blob/master/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java#L1579-L1588.

I'm working on a refactor https://github.com/apache/beam/pull/12710, which will provide much more test coverage. Sorry for the inconvenience.

Reply:
No worries. Please close this PR though if you don't think it's the best solution.